osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs

```diff
@@ -0,0 +1,364 @@
+        {
+            Default = -1
+        };
+
+        private static readonly BindableBool random_seed = new BindableBool
+        {
+            Value = true,
+            Default = true
+        };
+
+        [SettingSource("Random seed", "Generate a random seed for the beatmap generation")]
+        public BindableBool RandomSeed => random_seed;
+
+        [SettingSource("Seed", "Seed for the random beatmap generation", SettingControlType = typeof(OsuModRandomSettingsControl))]
+        public Bindable<int> Seed => seed;
+
+        internal static bool CustomSeedDisabled => random_seed.Value;
+
+        public OsuModRandom()
+        {
+            if (seed.Default != -1)
+                return;
+
+            var random = RNG.Next();
+            seed.Value = random;
+            seed.Default = random;
+            seed.BindValueChanged(e => seed.Default = e.NewValue);
+        }
+
+        public void ApplyToBeatmap(IBeatmap iBeatmap)
+        {
+            if (!(iBeatmap is OsuBeatmap beatmap))
+                return;
+
+            if (RandomSeed.Value)
+                seed.Value = RNG.Next();
+
+            var rng = new Random(seed.Value);
+
+            // Absolute angle
+            float prevAngleRad = 0;
+
+            // Absolute positions
+            Vector2 prevPosUnchanged = beatmap.HitObjects[0].Position;
+            Vector2 prevPosChanged = beatmap.HitObjects[0].Position;
+
+            // rateOfChangeMultiplier changes every i iterations to prevent shaky-line-shaped streams
+            byte i = 3;
+            float rateOfChangeMultiplier = 0;
+
+            foreach (var beatmapHitObject in beatmap.HitObjects)
+            {
+                if (!(beatmapHitObject is OsuHitObject hitObject))
+                    return;
+
+                // posUnchanged: position from the original beatmap (not randomised)
+                var posUnchanged = hitObject.EndPosition;
+                var posChanged = Vector2.Zero;
+
+                // Angle of the vector pointing from the last to the current hit object
+                float angleRad = 0;
+
+                if (i >= 3)
+                {
+                    i = 0;
+                    rateOfChangeMultiplier = (float)rng.NextDouble() * 2 - 1;
+                }
+
+                if (hitObject is HitCircle circle)
+                {
+                    var distanceToPrev = Vector2.Distance(posUnchanged, prevPosUnchanged);
+
+                    circle.Position = posChanged = getRandomisedPosition(
+                        rateOfChangeMultiplier,
+                        prevPosChanged,
+                        prevAngleRad,
+                        distanceToPrev,
+                        out angleRad
+                    );
+                }
+
+                // TODO: Implement slider position randomisation
+
+                prevAngleRad = angleRad;
+                prevPosUnchanged = posUnchanged;
+                prevPosChanged = posChanged;
```

Comment:
I'd probably bundle these three in a struct type that can be passed around. Would probably clean up `getRandomisedPosition()` too which is quite ugly due to number of params and returning stuff both via return value and `out` parameters.